### PR TITLE
proc-cmdline: reimplement proc_cmdline_get_key_many_internal()

### PR DESCRIPTION
**Pull request by yuwata**
No functional change, just refactoring.

Split-out of #26887.

### CODE REVIEW

File: src/basic/proc-cmdline.c [hunk 1]
```diff
@@ -303,75 +303,91 @@ int proc_cmdline_get_bool(const char *key, bool *ret) {
...
-                                p = line;
-                                continue;
-                        }
+        FOREACH_ARRAY(e, c->data, c->num) {
+                if (!proc_cmdline_key_streq(e->key, key))
+                        continue;
 
-                        break;
+                if (value) {
+                        r = free_and_strdup(&e->value, value);
```
Comment (keszybz) [inline src/basic/proc-cmdline.c:337]: Hmm, I would expect us to replace `e->value` unconditionally. I.e. drop the `if`.

File: src/basic/proc-cmdline.c [hunk 2]
```diff
@@ -303,75 +303,91 @@ int proc_cmdline_get_bool(const char *key, bool *ret) {
...
+                        if (r < 0)
+                                return r;
                 }
```
Comment (keszybz) [inline src/basic/proc-cmdline.c:340]: No `break`?

File: src/basic/proc-cmdline.c [hunk 3]
```diff
@@ -303,75 +303,91 @@ int proc_cmdline_get_bool(const char *key, bool *ret) {
...
-int proc_cmdline_get_key_many_internal(ProcCmdlineFlags flags, ...) {
-        _cleanup_free_ char *line = NULL;
-        bool processing_efi = true;
-        const char *p;
-        va_list ap;
-        int r, ret = 0;
+typedef struct ProcCmdlineDataEntry {
+        const char *key;
+        char *value;
+        char **ret;
```
Comment (keszybz) [inline src/basic/proc-cmdline.c:309]: What is `ret`? I think a more descriptive name would be better.

**Review by keszybz — COMMENTED** [review]
Hmm, on their own, I'm not sure if those changes are useful. This version is much longer and it also seems harder to grok, with different layers. In the old version, things are much more straightforward.

**Comment by yuwata** [comment]
> Hmm, on their own, I'm not sure if those changes are useful. This version is much longer and it also seems harder to grok, with different layers. In the old version, things are much more straightforward.

OK. I will take another way. Thank you for the comments. Closing.

**Pull request closed without merging**